### PR TITLE
Section Swap Backend

### PR DESCRIPTION
**Pull request by edwardneo**

### CODE REVIEW

**Comment by ericyche** [comment]
make wording more descriptive -1

**Comment by edwardneo** [comment]
:(

File: csm_web/scheduler/models.py
Comment (ericyche) [inline csm_web/scheduler/models.py:254]: lmao

Comment (KartavyaSharma) [reply csm_web/scheduler/models.py]: Fixed.

Comment (ericyche) [reply csm_web/scheduler/models.py]: thank you kurt :holding_back_tears:

File: csm_web/scheduler/tests/models/test_swap.py
Comment (smartspot2) [inline csm_web/scheduler/tests/models/test_swap.py:76]: This should be split up into two tests; the parameterization should be for a single _kind_ of test with various different inputs; here, you're testing two different situations; one where the student has no requests, and one where the student has a swap request.

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:830]: Generally you should never be catching the generic `Exception` object, as this means that any unexpected errors would also be caught and we wouldn't know about it. Make all of the `except Exception:` blocks throughout more specific, for _only_ the kinds of exceptions you expect could happen as a result of user error.

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:859]: User input errors should be returned as a 400 Bad Request error, not a 403 Permission Denied error (unsure if there's a builtin response class for this). A 403 (raised by `PermissionDenied`) is for when the user is not authorized to access a resource or perform an action.

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:945]: There's a bulk delete method for a `QuerySet`: https://docs.djangoproject.com/en/4.2/ref/models/querysets/#delete

Using a for loop sends separate queries for each delete action, while a delete on a query set will send one query to delete all of the objects in the query set.

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:976]: There's no need to save the object if you've just deleted it.

File: csm_web/scheduler/tests/models/test_swap.py
Comment (smartspot2) [inline csm_web/scheduler/tests/models/test_swap.py:233]: Add a newline at the end of the file >:(

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:903]: Since this inline conditional is quite long (it spans two lines), it's better convention (and more readable) to split it up into if/else blocks.

This should conventionally default to None if the field is not present, not -1. Additionally, the allowed request methods are only POST and DELETE, so the conditional doesn't quite make sense to me; the `else` part is never executed.

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:956]: typo

**Review by smartspot2 — CHANGES_REQUESTED** [review]
Overall, this looks pretty good; just some comments on convention. I haven't actually run any of the code though; I'll likely do that when this gets merged into the main feature PR, as when integration is done, some bugs are likely to surface.

**Comment by KartavyaSharma** [comment]
@smartspot2 Fixed all the requested changes, do you want me to merge this into the feature branch?

File: csm_web/scheduler/views/section.py
Comment (smartspot2) [inline csm_web/scheduler/views/section.py:38]: These seem to break the existing section tests due to the change in URL matching. I'm unfamiliar with these viewset fields, but this should be looked into; either modify this, or correct the existing tests (and potentially serializers?) that may rely on the URL match groups.

**Review by smartspot2 — CHANGES_REQUESTED** [review]
I've just pointed out a couple more small issues; I've also ran the pytests locally, and they seem to be failing due to the changes in the URL matching for the Section viewset. After you've fixed these, you can merge this into the main feature branch (it doesn't need approval).

When you've merged this in, make sure you also rebase the main feature branch onto master, and ensure that these changes work with the new Django version and with Docker. The CI should automatically run with that PR when this is merged in, so you can fix any additional bugs that may have popped up there too.

File: csm_web/scheduler/views/section.py
Comment (KartavyaSharma) [reply csm_web/scheduler/views/section.py]: @edwardneo I remember this coming from a commit of yours for the swap DELETE method. I have removed this for now since it's messing up the URL matcher for all the other tests and I was able to get the swap tests to work without this by making a few changes in the parameter names for the `swap_with_id` and `swap_no_id` functions. Lmk if you need me to add this back for any code extensions you plan to make in the future. Leaving this unresolved for now.

**Comment by KartavyaSharma** [comment]
@smartspot2 Should I squash + merge here or rebase + merge, what do we usually do for merging into feature branches?

**Pull request merged**
